tensorflow/core/kernels/unique_op_gpu.cu.cc

```diff
@@ -0,0 +1,459 @@
+    // 4) Sort the extracted unique input indices to put them in order of
+    //    first appearance.
+    //      sorted_unique_input_inds, sorted_unique_perm =
+    //          sort(unique_input_inds)
+    // sorted_unique_input_inds: [0, 1, 3, 4, 6, 7, 8, 11, 17]
+    // sorted_unique_perm: [2, 4, 3, 0, 8, 7, 5, 6, 1]
+    // 5) Gather the sorted unique input values to produce output, and invert
+    //    the second sort permutation to produce an inverse ID mapping. If
+    //    counts are required, also take the adjacent difference between
+    //    segment_ends indices to produce counts.
+    //      output = input[sorted_unique_input_inds]
+    //      inv_sorted_unique_perm[sorted_unique_perm[i]] = i
+    //      counts = adjacent_difference(segment_ends)
+    // output: [3, 5, 4, 1, 9, 8, 6, 7, 2]
+    // inv_sorted_unique_perm: [3, 8, 0, 2, 1, 6, 7, 5, 4]
+    // counts: [3, 3, 4, 1, 1, 3, 3, 1, 1]
+    // 6) Look up unique IDs via the inverse ID mapping and scatter them using
+    //    the original sort permutation to produce the indices output.
+    //      idx[sorted_input_inds] =
+    //          inv_sorted_unique_perm[sorted_input_unique_ids]
+    // idx: [0, 1, 0, 2, 3, 2, 4, 5, 6, 0, 1, 7, 5, 5, 2, 6, 2, 8, 1, 6]
+
+    Tensor sorted_input_inds;
+    TIndex* sorted_input_inds_ptr = nullptr;
+    AllocateTemp(context, input_size, &sorted_input_inds,
+                 &sorted_input_inds_ptr, done);
+    if (!context->status().ok()) return;
+
+    Tensor sorted_input;
+    T* sorted_input_ptr = nullptr;
+    AllocateTemp(context, input_size, &sorted_input, &sorted_input_ptr, done);
+    if (!context->status().ok()) return;
+
+    const T* input_ptr = input.flat<T>().data();
+    OP_REQUIRES_OK_ASYNC(
+        context,
+        (GpuRadixSort(context, input_size, /*keys_in=*/input_ptr,
+                      /*keys_out=*/sorted_input_ptr,
+                      /*indices_in=*/static_cast<const TIndex*>(nullptr),
+                      /*indices_out=*/sorted_input_inds_ptr)),
+        done);
+
+    // Create a fancy input iterator to indicate segment boundaries.
+    gpuprim::TransformInputIterator<bool, SegmentIndicatorFunctor<T, TIndex>,
+                                    gpuprim::CountingInputIterator<TIndex>>
+        segment_indicator_iter(0, {sorted_input_ptr});
+
+    Tensor sorted_input_unique_ids;
+    TIndex* sorted_input_unique_ids_ptr = nullptr;
+    AllocateTemp(context, input_size, &sorted_input_unique_ids,
+                 &sorted_input_unique_ids_ptr, done);
+    if (!context->status().ok()) return;
+
+    OP_REQUIRES_OK_ASYNC(
+        context,
+        GpuInclusivePrefixSum(context, input_size, segment_indicator_iter,
+                              sorted_input_unique_ids_ptr),
+        done);
+
+    // Copy the last element of sorted_input_unique_ids back to the host to
+    // obtain uniq_size.
+    ScratchSpace<TIndex> last_idx_host(context, 1, /*on_host=*/true);
+    if (input_size > 0) {
```

Comment:
Why not early exit if `input_size` is `0`?  (Please add a comment.)

Reply:
Added early exit instead.